### PR TITLE
TextEncoder/TextDecoder SharedArrayBuffer tests

### PR DESCRIPTION
**Pull request by Bnaya**
Add tests for TextEncoder/TextEncoder with SharedArrayBuffer

### CODE REVIEW

File: encoding/streams/decode-utf8.any.js [hunk 1]
```diff
@@ -40,22 +40,25 @@ promise_test(async () => {
...
-  const output = input.pipeThrough(new TextDecoderStream());
-  const array = await readableStreamToArray(output);
-  assert_array_equals(array, [], 'no chunks should be output');
-}, 'decoding a transferred ArrayBuffer chunk should give no output');
+[ArrayBuffer, SharedArrayBuffer].forEach((ArrayBufferOrSharedArrayBuffer) => {
+  promise_test(async () => {
+    const buffer = new ArrayBufferOrSharedArrayBuffer(3);
+    const view = new Uint8Array(buffer, 1, 1);
+    view[0] = 65;
+    new MessageChannel().port1.postMessage(buffer, [buffer]);
```
Comment (annevk) [inline encoding/streams/decode-utf8.any.js]: This test cannot be generalized this easily unfortunately. We'd also have to add headers and then we can no longer use `.any.js` syntax. I'm also not really sure if streams work with `SharedArrayBuffer`. @ricea?

Comment (ricea) [reply encoding/streams/decode-utf8.any.js]: Streams are agnostic about SharedArrayBuffer[1]. Up until now I have interpreted

> Values of the IDL buffer source types are represented by objects of the corresponding ECMAScript class, with the additional restriction that unless the type is associated with the [AllowShared] extended attribute, they can only be backed by ECMAScript ArrayBuffer objects, and not SharedArrayBuffer objects.

from the WebIDL standard to imply that

> Let bufferSource be the result of converting chunk to a BufferSource.

in the encoding standard means that we don't accept SharedArrayBuffers.

---

[1] Except for byte streams. But this standard doesn't currently use byte streams.

File: encoding/encodeInto.any.js [hunk 2]
```diff
@@ -73,77 +73,83 @@
...
-            buffer = new ArrayBuffer(bufferLength),
-            view = new Uint8Array(buffer, destinationOffset, destinationLength),
-            fullView = new Uint8Array(buffer),
-            control = new Array(bufferLength);
-      let byte = destinationFiller;
-      for (let i = 0; i < bufferLength; i++) {
-        if (destinationFiller === "random") {
-          byte = Math.floor(Math.random() * 256);
+  ].forEach((destinationData) => {
+    [ArrayBuffer, SharedArrayBuffer].forEach((ArrayBufferOrSharedArrayBuffer) => {
```
[the remark this answers is not in the document]
Comment (Bnaya) [reply encoding/encodeInto.any.js]: Done

File: encoding/encodeInto.any.js [hunk 3]
```diff
@@ -73,77 +73,83 @@
...
+     assert_equals(read, 0);
+     assert_equals(written, 0);
+     new MessageChannel().port1.postMessage(buffer, [buffer]);
+     ({ read, written } = new TextEncoder().encodeInto("", view));
+     assert_equals(read, 0);
+     assert_equals(written, 0);
+     ({ read, written } = new TextEncoder().encodeInto("test", view));
+     assert_equals(read, 0);
+     assert_equals(written, 0);
+   }, "encodeInto() and a detached output buffer(" + ArrayBufferOrSharedArrayBuffer.name + ")");
```
[the remark this answers is not in the document]
Comment (Bnaya) [reply encoding/encodeInto.any.js]: I took out the detached part

File: encoding/encodeInto.any.js [hunk 4]
```diff
@@ -73,77 +73,91 @@
...
-            buffer = new ArrayBuffer(bufferLength),
-            view = new Uint8Array(buffer, destinationOffset, destinationLength),
-            fullView = new Uint8Array(buffer),
-            control = new Array(bufferLength);
-      let byte = destinationFiller;
-      for (let i = 0; i < bufferLength; i++) {
-        if (destinationFiller === "random") {
-          byte = Math.floor(Math.random() * 256);
+  ].forEach((destinationData) => {
+    [ArrayBuffer, self["SharedArrayBuffer"]].forEach((ArrayBufferOrSharedArrayBuffer) => {
```
Comment (annevk) [inline encoding/encodeInto.any.js]: I guess this works, but it seems slightly better to use `"ArrayBuffer"` and `"SharedArrayBuffer"` here and then use `new self[ArrayBufferOrSharedArrayBuffer]` inside the test. That way there's not a bunch of undefined weirdness and instead the exception will make some sense.

File: encoding/encodeInto.any.js [hunk 5]
```diff
@@ -73,77 +73,91 @@
...
- Float64Array].forEach(view => {
-  test(() => {
-    assert_throws(new TypeError(), () => new TextEncoder().encodeInto("", new view(new ArrayBuffer(0))));
-  }, "Invalid encodeInto() destination: " + view.name);
- });
-
-test(() => {
-  assert_throws(new TypeError(), () => new TextEncoder().encodeInto("", new ArrayBuffer(10)));
-}, "Invalid encodeInto() destination: ArrayBuffer");
+[ArrayBuffer, self["SharedArrayBuffer"]].forEach((ArrayBufferOrSharedArrayBuffer) => {
```
Comment (annevk) [inline encoding/encodeInto.any.js]: You could also wrap this forEach around the first two tests avoiding the duplication of the typed arrays forEach.

**Review by annevk — COMMENTED** [review]
@hsivonen do you have any other tests you'd like to see added/modified? (Can also be done incrementally in a new PR perhaps.)

File: encoding/streams/decode-utf8.any.js (hunk 1 above)
Comment (annevk) [reply encoding/streams/decode-utf8.any.js]: I see, I guess we want to track addressing that separately or should we add SharedArrayBuffer support there at the same time?

File: encoding/encodeInto.any.js (hunk 5 above)
Comment (Bnaya) [reply encoding/encodeInto.any.js]: Done

File: encoding/encodeInto.any.js [hunk 6]
```diff
@@ -73,65 +73,74 @@
...
+        for (let i = 0; i < bufferLength; i++) {
+          if (i < destinationOffset || i >= (destinationOffset + testData.written.length)) {
+            assert_equals(fullView[i], control[i]);
+          } else {
+            assert_equals(fullView[i], testData.written[i - destinationOffset]);
+          }
         }
-      }
-    }, "encodeInto() with " + testData.input + " and destination length " + testData.destinationLength + ", offset " + destinationData.destinationOffset + ", filler " + destinationData.filler);
+      }, "encodeInto() into "  + self[arrayBufferOrSharedArrayBuffer].name + " with " + testData.input + " and destination length " + testData.destinationLength + ", offset " + destinationData.destinationOffset + ", filler " + destinationData.filler);
```
Comment (annevk) [inline encoding/encodeInto.any.js]: You can use `arrayBufferOrSharedArrayBuffer` on its own here now (and elsewhere).

Comment (Bnaya) [reply encoding/encodeInto.any.js]: Done!

File: encoding/streams/decode-utf8.any.js (hunk 1 above)
Comment (ricea) [reply encoding/streams/decode-utf8.any.js]: Either is fine with me. The changes should not be large. If we do it in this PR then we also need to change https://github.com/web-platform-tests/wpt/blob/master/encoding/streams/decode-bad-chunks.any.js and https://github.com/web-platform-tests/wpt/blob/master/encoding/streams/encode-bad-chunks.any.js which explicitly assert that SharedArrayBuffer doesn't work.

**Comment by hsivonen** [comment]
> do you have any other tests you'd like to see added/modified? (Can also be done incrementally in a new PR perhaps.)

I think it's enough to have `SharedArrayBuffer` testing with _some_ contents. At least the initial implementations will likely be indifferent to actual content.

If implementations ever want to do fancy optimizations, it would be good to test things like a stretch of more than 16 ASCII bytes followed by UTF-8 sequences with a variety of lead bytes followed by buffer end, but the kind of failure modes triggered would be very timing-dependent anyway and not really testable.

In summary, I think there's no need to add anything specific regarding buffer content at this point.

File: encoding/streams/decode-utf8.any.js (hunk 1 above)
Comment (annevk) [reply encoding/streams/decode-utf8.any.js]: @Bnaya are you interested in taking these on as well?

Comment (Bnaya) [reply encoding/streams/decode-utf8.any.js]: If i got it right:
need to remove the fail-test with SAB from here:
https://github.com/web-platform-tests/wpt/blob/master/encoding/streams/decode-bad-chunks.any.js

Not sure what do to in `encode-bad-chunks.any.js`

Comment (ricea) [reply encoding/streams/decode-utf8.any.js]: > If i got it right:
need to remove the fail-test with SAB from here:
https://github.com/web-platform-tests/wpt/blob/master/encoding/streams/decode-bad-chunks.any.js

Correct.

> Not sure what do to in `encode-bad-chunks.any.js`

Sorry, `encode-bad-chunks.any.js` doesn't need anything doing. My mistake.

Comment (annevk) [reply encoding/streams/decode-utf8.any.js]: I think we might also want to branch `decode-utf8.any.js` as you did before (although I can't tell exactly as that's overwritten now), but not the last two tests. That way we know for sure that `TextDecoderStream` also accepts `SharedArrayBuffer`.

Comment (Bnaya) [reply encoding/streams/decode-utf8.any.js]: Done, see the last commit

**Review by annevk — APPROVED** [review]
This looks good to me, module some style nits I may or may not care enough to fix myself. @ricea any feedback?

**Comment by ricea** [comment]
> This looks good to me, module some style nits I may or may not care enough to fix myself. @ricea any feedback?

Looks good.

**Comment by Bnaya** [comment]
If the repo had prettier, editor config, eslint or so,
That would have made contributing abit more friendly
